dtm0/doc/dtx.txt

```diff
@@ -0,0 +1,441 @@
+}
+
+ast_cb process_persistent_notice(fop, ...) {
+	struct dtm_service service = fop->service;
+
+	// Take a read lock on the container of records.
+	rw_read_lock(service->ds_log);
+
+	record = service->ds_log->find(fop->data);
+
+	// Take a write lock on the log entry.
+	rw_write_lock(record);
+
+	record->txr[fop->source.id].persistent = true;
+	
+	// is_stable = forall(i=0..N-1, record.txr[i].persistent == true)
+	if (!is_stable(record)) {
+		goto out;
+	}
+
+	dtx_sm_group = record->dtx->d_sm_group;
+
+	// We cannot directly modify m0 client data
+	// because we are running within DTM service SM group.
+	// We need a bridge between DTM service SM group and
+	// M0 Client SM group. Moreover, we need to change
+	// the state of DTX first, and then change the state
+	// of M0 client operation (and also we should call op_stable).
+
+	// Ast(DTM Service) -> Ast(DTX): a bridge between SM groups
+	// Ast(DTX) -> Clink(Client) -> Ast(Client): a bridge between
+        //                                           the components 
+	//                                           (Client and DTX).
+	// It gives us the following sequence:
+	//	Ast1 -> Ast2 -> Clink -> Ast3.
+	
+	ast post dtx_sm_group ast_on_dtx_stable(...);
+
+out:
+	unlock(record);
+	unlock(service->ds_log);
+}
+
+// The function is called by RPC when it gets a PERSISTENT notice
+// from another node.
+fop_process_persistent_notice(...) {
+	post ast dtm_service.sm_group process_persistent_notice(fop);
+}
+
+
+--------------------------------------------------------------------------------
+# NOTES
+
+CS should check the previous timestamp, and it should wait until the value
+changes. It could be a busy loop.
+
+We need to preserve the existing logic (that is not "DTM-aware") by adding
+conditions to check a compile-time flag or a run-time context. For example:
+
+// compile-time check
+void init_op(...) {
+#ifdef DTM0
+	op->dtx = init_dtx(...);
+#else
+	op->dtx = NULL;
+#endif
+}
+
+// run-time check
+void on_dix_reply_cb(...) {
+	if (op->dtx == NULL) {
+		handle_reply(op);
+	} else {
+		handle_reply_with_dtx(op);
+	}
+}
+
+Such conditions are required because of the following reasons:
+
+* Read requests are not executed within DTX but they still may go through
+  a code path used by DTM-aware code.
+
+* Meta operations should not use DTX at all.
+
+* Having the ability to switch between these two modes (with/without DTM) helps
+  with testing and debugging. Also, it makes easier to propagate the changes
+  into the upstream.
+
+--------------------------------------------------------------------------------
+# Q&A
+
+Q1: Is M0Client "clinked" with dixreq? or is it clinked with dtx.sm?
+A1: it should be linked with dtx.
+Q1.1: Should we extend the set of states of dixreq instead of adding dtx.sm?
+A1.1: We keep using dtx; no need to change dixreq states.
+Q2: Is Executed == got-one-reply or Executed == got-all-replies?
+A2: We assume a dtx is Executed when we got at least one reply.
```

Comment:
Generally, the result of such execution is known and returned to the dtx user... 
`at least one reply' is correct for current implementation but with spares or overwrites it may change. So it's better to use the first one.